### ISSUE TITLE
DefaultItemListImpl vs. multithread

### ISSUE THREAD

**Issue by RobbWatershed** [mikepenz/FastAdapter#959]
So I've been trying to understand the most frequent crashes Firebase reports for my app.

Here's a preview of a typical case, on every Android version and many phone models (302 similar crashes for 220 user in the last 24 hours) :

```
Fatal Exception: java.lang.IndexOutOfBoundsException: Index: 25, Size: 1
       at java.util.ArrayList.get(ArrayList.java:437)
       at com.mikepenz.fastadapter.utils.DefaultItemListImpl.get(DefaultItemListImpl.java:23)
       at com.mikepenz.fastadapter.adapters.ModelAdapter.getAdapterItem(ModelAdapter.java:168)
       at com.mikepenz.fastadapter.FastAdapter.getItem(FastAdapter.java:507)
       at com.mikepenz.fastadapter.FastAdapter.getItemViewType(FastAdapter.java:581)
       at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6186)
       at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6118)
       at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6114)
       at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2303)
       at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1627)
       at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1587)
       at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:665)
       at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4134)
       at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3851)
       at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4404)
       at android.view.View.layout(View.java:23753)
       at android.view.ViewGroup.layout(ViewGroup.java:7277)
       at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
       at android.view.View.layout(View.java:23753)
       at android.view.ViewGroup.layout(ViewGroup.java:7277)
       at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
       at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
       at android.view.View.layout(View.java:23753)
       at android.view.ViewGroup.layout(ViewGroup.java:7277)
       at androidx.recyclerview.widget.RecyclerView$LayoutManager.layoutDecoratedWithMargins(RecyclerView.java:9587)
       at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1685)
       at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1587)
       at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:665)
       at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4134)
       at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3851)
       at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4404)
       at android.view.View.layout(View.java:23753)
       at android.view.ViewGroup.layout(ViewGroup.java:7277)
       at androidx.viewpager2.widget.ViewPager2.onLayout(ViewPager2.java:527)
       at android.view.View.layout(View.java:23753)
       at android.view.ViewGroup.layout(ViewGroup.java:7277)
       at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
       at android.view.View.layout(View.java:23753)
       at android.view.ViewGroup.layout(ViewGroup.java:7277)
       at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
       at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
       at android.view.View.layout(View.java:23753)
       at android.view.ViewGroup.layout(ViewGroup.java:7277)
       at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
       at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
       at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
       at android.view.View.layout(View.java:23753)
```

The calling code is classic stuff
```java
compositeDisposable.add(Single.fromCallable(() -> FastAdapterDiffUtil.INSTANCE.calculateDiff(itemAdapter, contentItems))
                    .subscribeOn(Schedulers.computation())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(diffResult -> {
                        FastAdapterDiffUtil.INSTANCE.set(itemAdapter, diffResult);
                        differEndCallback();
                    })
            );
```

The more I'm thinking about this, the more I've come to believe this issue could be the consequence of `DefaultItemListImpl`'s list being used by different threads.

The following scenario is very likely in the case of my app -though I can't repro it- :
1. Initial list has 50 items
2. The list is updated by `LiveData` from 50 items to 1 item after a search query
3. Meanwhile, the displayed `RecyclerView` still tries to access item 25 (i.e. what we see in the stacktrace above)

In my opinion, this is likely to happen when the user browses his books very quickly on a slow device (where UI doesn't quite catch up with the flux of data from the DB).

My question is as follows : what would be the best approach to fix that problem ?

I'd suggest the following changes on `DefaultItemListImpl` :
- Use `Collections.synchronizedList()` on `DefaultItemListImpl._items` and use `synchronized` on every class method that manipulates the list
- Perform a simple range test on `get` and `set` to make certain we don't access stuff outside the list range

What's your take about these ? 😄

**Comment by mikepenz** [comment]
That's a touch one.

Ultimately it comes down to the target's application implementation to handle it carefully. Updating the RV form multiple thread would equally lead to exceptions. 

Just handling these and catching the problem would hide potential issues which would be super hard to debug. 


Would it be an option in your case to only provide the updates to the UI on the main thread?

The computation of diffs occurs async, and then in sync the updates will be dispatched to the RV. it should be handled so that the async diff computation don't happen in parallel though to also ensure it won't compute diffs based on wrong values :/ 


----

Performing the range test won't ensure that you get the item it expects to get, possibly causing other issues which will be kinda impossible to debug.

**Comment by RobbWatershed** [comment]
I'll give it a try for my next update.

Considering the volume of items to compare (hundreds of them, for some users), there's a good chance some people will report UI slowdowns. I'll keep you informed.

**Comment by mikepenz** [comment]
@RobbWatershed hmmm I would assume the time consuming part is computing the diff. not dispatching it to the RV. :/ but I may be wrong.

**Comment by RobbWatershed** [comment]
Going back on the main thread by simply calling `FastAdapterDiffUtil.INSTANCE.set(itemAdapter, contentItems);` seems to have lowered the occurences of crashes, so we're definitely progressing 😁 

However, I still have weird crashes in the activities that are using `PagedModelAdapter`, exclusively on Samsung devices with Android 10 : 

`Fatal Exception: java.lang.IllegalStateException: Two different ViewHolders have the same stable ID. Stable IDs in your adapter MUST BE unique and SHOULD NOT change.
 ViewHolder 1:ContentViewHolder{d24334d position=6 id=-2106877327, oldPos=-1, pLpos:-1} 
 View Holder 2:ContentViewHolder{fc4ef13 position=5 id=-2106877327, oldPos=-1, pLpos:-1} androidx.recyclerview.widget.RecyclerView{fe7ddb0 VFED..... ......ID 0,0-1080,2023 #7f0901bb app:id/library_list}, adapter:com.mikepenz.fastadapter.FastAdapter@1653429, layout:androidx.recyclerview.widget.LinearLayoutManager@eb1e3ae, context:me.devsaki.hentoid.activities.LibraryActivity@a5c89fd
       at androidx.recyclerview.widget.RecyclerView.handleMissingPreInfoForChangeError(RecyclerView.java:4268)
       at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep3(RecyclerView.java:4192)`

I'm trying to get more info out of these by putting extra traces. Still not sure why. I'll keep you informed and eventually close the issue if it turns out to be unrelated.

**Comment by mikepenz** [comment]
Very interesting case indeed. :/ also it seems weird to only occur on samsung phones

**Comment by deepanshu23** [comment]
@mikepenz 
Hi, we are also getting this crash a lot
Is it possible instead of updating the list from the calculateDiff method we can let the caller update the list.
This will ensure that the list can be updated on main thread only
```kotlin
fun <A : ModelAdapter<Model, Item>, Model, Item : GenericItem> calculateDiff(adapter: A, items: List<Item>, callback: DiffCallback<Item> = DiffCallbackImpl(), detectMoves: Boolean = true): DiffUtil.DiffResult {
        //remember the old items
        val oldItems = prepare(adapter, items)
        val adapterItems = adapter.adapterItems

        //pass in the oldItem list copy as we will update the one in the adapter itself
        val result = DiffUtil.calculateDiff(FastAdapterCallback(oldItems, items, callback), detectMoves)

        //make sure the new items list is not a reference of the already mItems list
        postCalculate(adapter, items) -> **can we remove this step**

        return result
    }
```

**Comment by RobbWatershed** [comment]
Thanks for the input @deepanshu23 I thought there was something I was doing wrong.

On my side, I have gathered the following info, but haven't progressed in finding the core issue.

- When I said it happens on Samsung / Android 10, my view on data was biased. It actually happens on a variety of devices and also on Android 7 to 11. As most of my users have Samsung phones with Android 10, I was at first focused on that one report that has the strongest impact on my stats.

- On some devices, there seems to be a way to repro it by manipulating `SelectExtension`. However, it crashes much more frequently than on my release builds, and not on my debug builds, which makes everything harder to analyze

- This is not linked by any means to `PagedModelAdapter` and happens indistinctly using paging or not

**Comment by RobbWatershed** [comment]
@deepanshu23 I'm having another lead on the issue : would you happen to feed data to `FastAdapter` using `LiveData`, and more precisely, `MediatorLiveData` ?  If so, we gotta have a discussion about that.

**Comment by deepanshu23** [comment]
@RobbWatershed No I am not using LiveData
I know the issue in my code that in some places we are adding data to adapter directly and in some places we are using diffutil for that
IMO, postCalculate(adapter, items) step shouldn't be there, the adapter list must be only updated by the caller on main thread
From your code pasted above, may i know
1) Is your list only updated by this method
2) how are you cancelling the previous executing diffutil if some new data arrive

And lastly have you tried using your own diffutil and updating the data yourself

```kotlin
object MyDiffUtil {

    private val compositeDisposable = CompositeDisposable()

    @SuppressLint("CheckResult")
    fun <A : ModelAdapter<Model, Item>, Model, Item : GenericItem> applyDiffFUtil(
        adapter: A,
        newItems: List<Item>
    ) {
        compositeDisposable.clear() // this is important , you can use rxjava swithcMap as well

        val items = adapter.adapterItems.toList()
        compositeDisposable.add(
                Single.fromCallable { calculateDiffResult(items, newItems) }
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ diffResult ->
                    adapter.adapterItems.clear()
                    adapter.adapterItems.addAll(newItems)
                    FastAdapterDiffUtil.INSTANCE.set(itemAdapter, diffResult);
                            differEndCallback();
                })
            )
    }

    private fun <Item : GenericItem> calculateDiffResult(
        items: List<Item>,
        newItems: List<Item>
    ): DiffUtil.DiffResult {
        return DiffUtil.calculateDiff(object : DiffUtil.Callback() {

            override fun getOldListSize() = items.size

            override fun getNewListSize() = newItems.size

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                items[oldItemPosition].identifier == newItems[newItemPosition].identifier

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return items[oldItemPosition] == newItems[newItemPosition]
            }                
    }

}
```

**Comment by RobbWatershed** [comment]
@deepanshu23 
> Is your list only updated by this method

Yes it is

> how are you cancelling the previous executing diffutil if some new data arrive

That's something I haven't focused until now, but that specific issue is now my main focus.
Do you have any recommendation / tried method on how to do that ?

> lastly have you tried using your own diffutil and updating the data yourself

I did, but just by creating my own `AsyncDifferConfig.Builder`. I'm pretty sure `FastAdapter` does send list updates from the main thread.

***

My turn for a small question about your code : you wrote the following instructions side by side

```kotlin
adapter.adapterItems.clear() 
adapter.adapterItems.addAll(newItems) // self updating adapter on main thread
```

and

```kotlin
FastAdapterDiffUtil.INSTANCE.set(itemAdapter, diffResult);
```

But aren't they doing the exact same thing ? I thought the point of using `FastAdapterDiffUtil` was to avoid using `ItemAdapter.adapterItems.addAll`...

**Comment by RobbWatershed** [comment]
I'm documenting the results of my research here as it progresses.

> how are you cancelling the previous executing diffutil if some new data arrive

Found something new today : as I said, the crash happens both with paging and without paging.

By reviewing the code of Android's paging library, I found out that, when updating the adapter by calling `PagedModelAdapter.submitList`, it eventually arrives in AndroidX's `AsyncPagedListDiffer`, which happens to have a mechanism that ignores any executing diffutil but the last one that has been submitted (see comment line 179 : `incrementing generation means any currently-running diffs are discarded when they finish`)

What's more, near the end of this method (lines 330 and after), we can clearly see that the last diff is used to update the list _on the main thread_

```java
final PagedList<T> oldSnapshot = mSnapshot;
        final PagedList<T> newSnapshot = (PagedList<T>) pagedList.snapshot();
        mConfig.getBackgroundThreadExecutor().execute(new Runnable() {
            @Override
            public void run() {
                final DiffUtil.DiffResult result;
                result = PagedStorageDiffHelper.computeDiff(
                        oldSnapshot.mStorage,
                        newSnapshot.mStorage,
                        mConfig.getDiffCallback());

                mMainThreadExecutor.execute(new Runnable() { // <-- here
                    @Override
                    public void run() {
                        if (mMaxScheduledGeneration == runGeneration) {
                            latchPagedList(pagedList, newSnapshot, result,
                                    oldSnapshot.mLastLoad, commitCallback);
                        }
                    }
                });
            }
        });
```

As promising as this theory was, I have now evidence that the crash we're talking about is not related to concurrent diffutils being mishandled, as they are handled properly when using paging, and it still produces crashes.

**Comment by RobbWatershed** [comment]
Okay, throw me stones if you wish, but I fear the issue lies in two single words : **hash collision**

I gotta confirm that by looking closer but I think the issue might be on my side, as I've been relying blindly on `Objects.hash` to provide unique identifiers, which seems to be a very bad idea.

Might be closing that issue soon 😉

**Comment by boguszpawlowski** [comment]
@RobbWatershed @mikepenz 
I believe I have found a way to reproduce it on FastAdapter's sample app (with some modifications in the code). Basically what I have done is loading a different amount of items with 100 milliseconds interval:
```
 private fun setDataAsync() {
        MainScope().launch {
            while (true) {
                delay(100)
                val data = createData()

                try {
                    withContext(Dispatchers.Default) {
                        val result =
                            FastAdapterDiffUtil.calculateDiff(fastItemAdapter.itemAdapter, data)
                        withContext(Dispatchers.Main) {
                            FastAdapterDiffUtil[fastItemAdapter.itemAdapter] = result
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }
```	
You can check complete code in here: https://github.com/boguszpawlowski/FastAdapter/blob/diffutil_bug/app/src/main/java/com/mikepenz/fastadapter/app/DiffUtilActivity.kt
I'm only starting the loop once, so there is no parallel modifications to the adapter itself. 
Now, if I start to interact with the screen (e.g. scrolling), I can get the crash with almost 100% reproducibility (note that it may take some time to get it, because you have to hit exact timing). 
My observations:
1. Issue does not occur when there is no interaction with the screen
2. Issue also does not occur without switching threads for diff calculation. 

So, I suspect that if some part of data is removed, and the user interact with the screen (or some change in view hierarchy happens), due to 2 different threads reading and modifying the list of items, crash can happen.

**Comment by RobbWatershed** [comment]
Well, I was about to announce my own problem has been actually solved (needed some fresh production stats to confirm that).

The issue has indeed been on my side all along : computing unique IDs using `Objects.hash` generated collisions regularly.
I switched to https://github.com/whitfin/siphash-java and that got rid of the issue !

=> As far as I'm concerned, the issue can be closed

**Comment by mikepenz** [comment]
That's awesome news. @RobbWatershed 

Really great investigation! and glad you were able to resolve it. 

Marked as `good to know` and will close :)

**Comment by boguszpawlowski** [comment]
Hmmm, then the crash that I found is different from yours @RobbWatershed, since in my app we are not using hash as unique id, nor does FastAdapter's sample app. I will add separate issue for it.
